### ISSUE TITLE
qp_functions_in_with test hangs frequently in CI when run in parallel_group

### ISSUE THREAD

**Issue by dhanashreek89** [greenplum-db/gpdb#4020]
### Greenplum version or build
Latest Master
```
 PostgreSQL 8.4devel (Greenplum Database 6.0.0-alpha.0+dev.1588.g70d0b90 build dev) on x86_64-apple-darwin16.4.0, compiled by GCC Apple LLVM version 8.0.0 (clang-800.0.42.1) compiled on Nov 29 2017 15:02:33 (with assert checking)
```
### OS version and uname -a
```
Darwin verano 16.4.0 Darwin Kernel Version 16.4.0: Thu Dec 22 22:53:21 PST 2016; root:xnu-3789.41.3~3/RELEASE_X86_64 x86_64
```
### Expected behavior
Test should finish successfully. More precisely the repro query should error out.
### Actual behavior
The query gets stuck.

### Step to reproduce the behavior
Recently we observed several occurrences of `qp_functions_in_with` test hanging in [gpdb_master ci](https://gpdb.data.pivotal.ci/teams/main/pipelines/gpdb_master/jobs/icw_gporca_centos7/builds/307) 
Query that is causing a hang:
```
CREATE TABLE foo (a int, b int);
INSERT INTO foo select i, i+1 from generate_series(1,10) i;
CREATE TABLE bar (c int, d int);
INSERT INTO bar select i, i+1 from generate_series(1,10) i;
CREATE FUNCTION func1_read_setint_sql_stb(x int) RETURNS setof int AS $$
DECLARE
    r int;
BEGIN
    FOR r in SELECT d FROM bar WHERE c <> $1
    LOOP
        RETURN NEXT r;
    END LOOP;
    RETURN;
END
$$ LANGUAGE plpgsql STABLE READS SQL DATA;

-- Query that gets stuck
WITH v(a, b) AS (SELECT func1_read_setint_sql_stb(a), b FROM foo WHERE b < 5) SELECT v1.a, v2.b FROM v AS v1, v AS v2 WHERE v1.a < v2.a order by v1.a, v2.b;

-- Plan Produced for the above query with ORCA:
explain WITH v(a, b) AS (SELECT func1_read_setint_sql_stb(a), b FROM foo WHERE b < 5) SELECT v1.a, v2.b FROM v AS v1, v AS v2 WHERE v1.a < v2.a order by v1.a, v2.b;
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1324463.79 rows=4 width=8)
   Merge Key: share0_ref2.func1_read_setint_sql_stb, share0_ref3.b
   ->  Sort  (cost=0.00..1324463.79 rows=2 width=8)
         Sort Key: share0_ref2.func1_read_setint_sql_stb, share0_ref3.b
         ->  Sequence  (cost=0.00..1324463.79 rows=2 width=8)
               ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=2 width=1)
                     ->  Materialize  (cost=0.00..431.00 rows=2 width=1)
                           ->  Result  (cost=0.00..431.00 rows=2 width=8)
                                 ->  Table Scan on foo  (cost=0.00..431.00 rows=2 width=8)
                                       Filter: b < 5
               ->  Nested Loop  (cost=0.00..1324032.79 rows=2 width=8)
                     Join Filter: share0_ref2.func1_read_setint_sql_stb < share0_ref3.func1_read_setint_sql_stb
                     ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=2 width=8)
                     ->  Materialize  (cost=0.00..431.00 rows=4 width=4)
                           ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=4 width=4)
                                 ->  Result  (cost=0.00..431.00 rows=2 width=4)
                                       ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=2 width=4)
 Optimizer status: PQO version 2.50.2
(18 rows)
```
When the hang occurs, Shared Scan reader is stuck in `shareinput_reader_waitready()`, polling the file descriptor for availability, using "select":
```
#0  0x00007f79b690d7a3 in select () from /lib64/libc.so.6
#1  0x000000000072bd0b in shareinput_reader_waitready (share_id=0, planGen=PLANGEN_OPTIMIZER) at nodeShareInputScan.c:629
#2  0x000000000072c11a in ExecSliceDependencyShareInputScan (node=node@entry=0x1d1bc18) at nodeShareInputScan.c:301
#3  0x00000000006ead88 in ExecSliceDependencyNode (node=<optimized out>) at execProcnode.c:978
#4  0x00000000006eaccd in ExecSliceDependencyNode (node=<optimized out>) at execProcnode.c:1002
#5  0x00000000006eaccd in ExecSliceDependencyNode (node=node@entry=0x1ccae10) at execProcnode.c:1002
#6  0x00000000006e7b98 in ExecutePlan (estate=estate@entry=0x1cc9440, planstate=planstate@entry=0x1ccae10, operation=operation@entry=CMD_SELECT,
    numberTuples=numberTuples@entry=0, direction=direction@entry=ForwardScanDirection, dest=dest@entry=0x1d31410) at execMain.c:2851
#7  0x00000000006e8543 in standard_ExecutorRun (queryDesc=0x1cdf5e0, direction=ForwardScanDirection, count=0) at execMain.c:935
#8  0x00000000008bb767 in PortalRunSelect (portal=0x1d043d0, forward=<optimized out>, count=0, dest=0x1d31410) at pquery.c:1140
#9  0x00000000008bd7e4 in PortalRun (portal=portal@entry=0x1d043d0, count=count@entry=9223372036854775807, isTopLevel=isTopLevel@entry=1 '\001',
    dest=dest@entry=0x1d31410, altdest=altdest@entry=0x1d31410, completionTag=completionTag@entry=0x7ffd96fe61d0 "") at pquery.c:961
#10 0x00000000008b52fd in exec_mpp_query (
    query_string=query_string@entry=0x1cf0382 "WITH v(a, b) AS (SELECT func1_read_setint_sql_stb(a), b FROM foo WHERE b < 5) SELECT v1.a, v2.b FROM v AS v1, v AS v2 WHERE v1.a < v2.a order by v1.a, v2.b;", serializedQuerytree=serializedQuerytree@entry=0x0,
    serializedQuerytreelen=serializedQuerytreelen@entry=0, serializedPlantree=serializedPlantree@entry=0x1cf041f "P\016",
    serializedPlantreelen=serializedPlantreelen@entry=743, serializedParams=serializedParams@entry=0x0, serializedParamslen=0,
    serializedQueryDispatchDesc=0x1cf0706 "_\001", serializedQueryDispatchDesclen=163, seqServerHost=0x1cf07a9 "10.254.0.78", seqServerPort=60287,
    localSlice=1) at postgres.c:1367
#11 0x00000000008bab20 in PostgresMain (argc=<optimized out>, argv=argv@entry=0x1b5db88, dbname=<optimized out>, username=<optimized out>)
    at postgres.c:5187
#12 0x000000000084a815 in BackendRun (port=0x1b58240) at postmaster.c:6698
#13 BackendStartup (port=0x1b58240) at postmaster.c:6381
#14 ServerLoop () at postmaster.c:2449
#15 0x000000000084cff2 in PostmasterMain (argc=argc@entry=12, argv=argv@entry=0x1b2dee0) at postmaster.c:1534
#16 0x00000000004c9c61 in main (argc=12, argv=0x1b2dee0) at main.c:206
(gdb)
```
The hang can be reproduced semi-consistently by busy-waiting in the middle of `secure_read`. It seems that the `client_read_ended()` function will clear pending cancel requests, which introduces a difficult-to-reproduce race.

Applying the following patch on an assert enable build and then running the above query with `set gp_loop_a_long_time=on` will reproduce the hang (roughly one in four/five times of query rerun):
```
diff --git a/src/backend/libpq/be-secure.c b/src/backend/libpq/be-secure.c
index 314bcce4e2..96ad91c57c 100644
--- a/src/backend/libpq/be-secure.c
+++ b/src/backend/libpq/be-secure.c
@@ -305,6 +305,9 @@ rloop:
                prepare_for_client_read();

                n = recv(port->sock, ptr, len, 0);
+               if (gp_loop_a_long_time) {
+                       for (int i = 0; i < 1000000000; ++i);
+               }

                client_read_ended();
        }
diff --git a/src/backend/utils/misc/guc_gp.c b/src/backend/utils/misc/guc_gp.c
index 6287273d0e..12d54f1c3b 100644
--- a/src/backend/utils/misc/guc_gp.c
+++ b/src/backend/utils/misc/guc_gp.c
@@ -562,6 +562,9 @@ int                 writable_external_table_bufsize = 64;
 bool           gp_enable_mk_sort = true;
 bool           gp_enable_motion_mk_sort = true;

+/* XXX */
+bool           gp_loop_a_long_time = false;
+
 static const struct config_enum_entry gp_log_format_options[] = {
        {"text", 0},
        {"csv", 1},
@@ -3381,6 +3384,15 @@ struct config_bool ConfigureNamesBool_gp[] =
                &gp_enable_segment_copy_checking,
                true, NULL, NULL
        },
+       {
+               {"gp_loop_a_long_time", PGC_USERSET, CUSTOM_OPTIONS,
+                       gettext_noop("XXX"),
+                       NULL,
+                       GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE | GUC_GPDB_ADDOPT
+               },
+               &gp_loop_a_long_time,
+               false, NULL, NULL
+       },
        /* End-of-list marker */
        {
                {NULL, 0, 0, NULL, NULL}, NULL, false, NULL, NULL
diff --git a/src/include/postgres.h b/src/include/postgres.h
index 66dbeb9e35..24aa75f968 100644
--- a/src/include/postgres.h
+++ b/src/include/postgres.h
@@ -482,6 +482,9 @@ extern int ExceptionalCondition(const char *conditionName,
                                         const char *errorType,
                                         const char *fileName, int lineNumber);

+/* XXX */
+extern bool gp_loop_a_long_time;
+

 #ifdef __cplusplus
 }   /* extern "C" */
```

### What is the issue?
- Shared Scan writer (executing in slice 2) starts and soon it errors out while executing `Result` that call to, `func1_read_setint_sql_stb()` which accesses a distributed relation `bar`.  This is not supported.
The error reported by writer is:
`ERROR:  function cannot execute on segment because it accesses relation "public.bar"  (seg0 slice2 127.0.0.1:25435 pid=61210)`
- This sends a Query Cancel Message to master and master now has to notify the Shared Scan reader (executing in slice 1).
- However if the timing is such that the writer is quick in reporting the error and master is quick enough in sending the signal to reader but the reader process has not yet started (because of OS scheduling/ heavy load); the signal gets cleared by `client_read_ended()`.

Hence If the signal is received before `client_read_ended()` is executed, the reader waits forever for writer (which has already errored out).

This seems like a general issue with the architecture and not specific to Shared Scan; this can occur with any other operator.

Upstream has a [patch](https://github.com/postgres/postgres/commit/82170c747b
) that on the surface seems to fix an issue similar to this.

However cherry-picking this upstream commit does not solve the problem and causes undesirable failures.

**Comment by vraghavan78** [comment]
@pengzhout @yydzero I think ur fix https://github.com/greenplum-db/gpdb/pull/4021 may be able to fix this. Thought?

**Comment by dhanashreek89** [comment]
Verified that repro query does not hang with the fix #4021. 
This issue can be closed once the PR gets merged.

**Comment by hlinnaka** [comment]
It's now been merged, so closing.
